src/server/routes/actions.spec.ts

```diff
@@ -0,0 +1,279 @@
+    expect(res.body.error).toBe(parameterError("address"));
+  });
+
+  describe("web endpoints", () => {
+    beforeEach(() => {
+      mockConfig.mockImplementation((type: string) => {
+        if (type === "EXTERNAL_ACCESS") {
+          return true;
+        }
+        return "generic";
+      });
+    });
+
+    test("should fail with no captcha", async () => {
+      const res = await request(app).post("/drip").send({ address: "example" });
+      expect(res.body.error).toBe(parameterError("recaptcha"));
+    });
+
+    test("should request drip", async () => {
+      mockHandleRequest.mockImplementation(() => {
+        return {};
+      });
+
+      const res = await request(app).post("/drip").send({ address: "example1", recaptcha: "captcha1" });
+      expect(mockHandleRequest).toHaveBeenCalledWith(
+        expect.objectContaining({ external: true, address: "example1", recaptcha: "captcha1" }),
+      );
+      expect(res.status).toBe(200);
+    });
+
+    test("should report error on error drip result", async () => {
+      const error = "this is an error in web";
+      mockHandleRequest.mockImplementation(() => {
+        return { error };
+      });
+
+      const res = await request(app).post("/drip").send({ address: "example2", recaptcha: "captcha2" });
+      expect(mockHandleRequest).toHaveBeenCalledWith(
+        expect.objectContaining({ external: true, address: "example2", recaptcha: "captcha2" }),
+      );
+      expect(res.body.error).toBe(error);
+    });
+
+    test("should report error on internal error", async () => {
+      mockHandleRequest.mockRejectedValueOnce("random error in web");
+      const res = await request(app).post("/drip").send({ address: "example3", recaptcha: "captcha3" });
+      expect(res.body.error).toBe("Operation failed.");
+      expect(res.status).toBe(400);
+      expect(mockLoggerError).toHaveBeenCalledWith("random error in web");
+    });
+  });
+
+  describe("bot endpoints", () => {
+    beforeEach(() => {
+      mockConfig.mockImplementation((type: string) => {
+        if (type === "EXTERNAL_ACCESS") {
+          return false;
+        }
+        return "generic";
+      });
+    });
+
+    test("should fail with no address", async () => {
+      const res = await request(app).post("/drip");
+      expect(res.body.error).toBe(parameterError("address"));
+      expect(res.status).toBe(400);
+    });
+
+    test("should fail with no amount", async () => {
+      const res = await request(app).post("/drip").send({ address: "example" });
+      expect(res.body.error).toBe(parameterError("amount"));
+      expect(res.status).toBe(400);
+    });
+
+    test("should fail with no sender", async () => {
+      const res = await request(app).post("/drip").send({ address: "example", amount: "100" });
+      expect(res.body.error).toBe(parameterError("sender"));
+      expect(res.status).toBe(400);
+    });
+
+    test("should fail with no sender", async () => {
```

Comment:
Copy pasted name?

Reply:
Yup, and it seems I also had the same problem in the previous test scope.